cmake/modules/extensions.cmake

```diff
@@ -1283,22 +1285,85 @@ endfunction(zephyr_linker_sources)
+# name to relocate.
+#
+# The FILES directive will relocate a list of files (wildcards supported)
+# This directive will relocate file1. and file2.c to SRAM:
+# zephyr_code_relocate(FILES file1.c file2.c LOCATION SRAM)
+# Note, files can also be passed as a comma separated list to support using
+# cmake generator arguments
+#
+# The LIBRARY directive will relocate a library
+# This directive will relocate the target my_lib to SRAM:
+# zephyr_code_relocate(LIBRARY my_lib SRAM)
+#
+# The following optional arguments are supported:
+# - NOCOPY: this flag indicates that the file data does not need to be copied
+#   at boot time (For example, for flash XIP).
+# - PHDR [program_header]: add program header. Used on Xtensa platforms.
+function(zephyr_code_relocate)
   set(options NOCOPY)
-  cmake_parse_arguments(CODE_REL "${options}" "" "" ${ARGN})
-  if(NOT IS_ABSOLUTE ${file})
-    set(file ${CMAKE_CURRENT_SOURCE_DIR}/${file})
+  set(single_args LIBRARY LOCATION PHDR)
+  set(multi_args FILES)
+  cmake_parse_arguments(CODE_REL "${options}" "${single_args}"
+    "${multi_args}" ${ARGN})
+  # Argument validation
+  if(CODE_REL_UNPARSED_ARGUMENTS)
+    message(FATAL_ERROR "zephyr_code_relocate(${ARGV0} ...) "
+      "given unknown arguments: ${CODE_REL_UNPARSED_ARGUMENTS}")
+  endif()
+  if((NOT CODE_REL_FILES) AND (NOT CODE_REL_LIBRARY))
+    message(FATAL_ERROR
+      "zephyr_code_relocate() requires either FILES or LIBRARY be provided")
+  endif()
+  if(CODE_REL_FILES AND CODE_REL_LIBRARY)
+    message(FATAL_ERROR "zephyr_code_relocate() only accepts "
+      "one argument between FILES and LIBRARY")
+  endif()
+  if(NOT CODE_REL_LOCATION)
+    message(FATAL_ERROR "zephyr_code_relocate() requires a LOCATION argument")
+  endif()
+  if(CODE_REL_LIBRARY)
+    # Use cmake generator expression to convert library to file list
+    set(genex_src_dir "$<TARGET_PROPERTY:${CODE_REL_LIBRARY},SOURCE_DIR>")
+    set(genex_src_list "$<TARGET_PROPERTY:${CODE_REL_LIBRARY},SOURCES>")
+    set(file_list
+      "${genex_src_dir}/$<JOIN:${genex_src_list},$<SEMICOLON>${genex_src_dir}/>")
+  else()
+    # Check if CODE_REL_FILES is a generator expression, if so leave it
+    # untouched.
+    string(GENEX_STRIP "${CODE_REL_FILES}" no_genex)
+    if(CODE_REL_FILES STREQUAL no_genex)
+      # no generator expression in CODE_REL_FILES, check if list of files
+      # is absolute
+      foreach(file ${CODE_REL_FILES})
+        if(NOT IS_ABSOLUTE ${file})
+          set(file ${CMAKE_CURRENT_SOURCE_DIR}/${file})
+        endif()
+        list(APPEND file_list ${file})
+      endforeach()
+    else()
+      # Generator expression is present in file list. Leave the list untouched.
+      set(file_list ${CODE_REL_FILES})
+    endif()
   endif()
   if(NOT CODE_REL_NOCOPY)
     set(copy_flag COPY)
   else()
     set(copy_flag NOCOPY)
   endif()
+  if(CODE_REL_PHDR)
+    set(CODE_REL_LOCATION "${CODE_REL_LOCATION}\ :${CODE_REL_PHDR}")
```

Comment:
exactly why is an escaped space needed here ?
That doesn't seem to have been necessary before.

Reply:
The PHDR directive was previously paired with the relocation location, as you can see in this diff: https://github.com/zephyrproject-rtos/zephyr/pull/50791/files#diff-7087c4d7b0a6595b501520d5eb10ed5ecb01b3540831e3a141a6a662632ac940R20. This directive makes sure that the PHDR is still paired with the code location, so it can be parsed by `gen_relocate_app.py`. `gen_relocate_app.py` treats the PHDR as being paired with the relocation location, so this line preserves that behavior with the new API. An alternative here would be to integrate PHDR as a parameter like NOCOPY is used.